sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PrivateFieldUsedAsLocalVariable.cs

```diff
@@ -36,340 +36,216 @@ namespace SonarAnalyzer.Rules.CSharp
 
-                    var classSymbol = c.SemanticModel.GetDeclaredSymbol(classNode);
-                    if (classSymbol == null)
+                    if (typeDeclaration.Modifiers.Any(SyntaxKind.PartialKeyword))
                     {
                         return;
                     }
 
-                    var classMethods = classSymbol.GetMembers().Where(m => m.Kind == SymbolKind.Method).ToHashSet();
+                    var privateFields = GetPrivateFields(c.SemanticModel, typeDeclaration);
 
-                    ExcludePrivateFieldsBasedOnReferences(privateFields, referencesByEnclosingSymbol, classMethods);
-                    ExcludePrivateFieldsBasedOnCompilerErrors(privateFields, referencesByEnclosingSymbol, classMethods,
-                        classNode, c.SemanticModel.Compilation);
+                    var collector = new FieldAccessCollector(c.SemanticModel, privateFields);
+                    collector.Visit(typeDeclaration);
 
-                    foreach (var privateField in privateFields.Values.Where(f => !f.Excluded))
-                    {
-                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, privateField.Syntax.GetLocation(),
-                            privateField.Symbol.Name));
-                    }
+                    privateFields.Keys
+                        .Where(collector.IsRemovableField)
+                        .ToList()
+                        .ForEach(ReportField);
+
+                    void ReportField(ISymbol fieldSymbol) =>
+                        c.ReportDiagnosticWhenActive(
+                            Diagnostic.Create(rule, privateFields[fieldSymbol].GetLocation(), fieldSymbol.Name));
                 },
-                SyntaxKind.ClassDeclaration, SyntaxKind.StructDeclaration);
+                SyntaxKind.ClassDeclaration,
+                SyntaxKind.StructDeclaration);
         }
 
-        private static IImmutableDictionary<ISymbol, PrivateField> GetPrivateFields(TypeDeclarationSyntax classNode,
-            SemanticModel semanticModel)
+        private static IDictionary<ISymbol, VariableDeclaratorSyntax> GetPrivateFields(SemanticModel semanticModel,
+            TypeDeclarationSyntax typeDeclaration)
         {
-            return classNode.Members
-                        .Where(m => m.IsKind(SyntaxKind.FieldDeclaration))
-                        .Cast<FieldDeclarationSyntax>()
-                        .Where(f => !f.AttributeLists.Any())
-                        .SelectMany(f => f.Declaration.Variables.Select(
-                            v => new PrivateField(v, semanticModel.GetDeclaredSymbol(v))))
-                        .Where(f => f.Symbol != null && f.Symbol.DeclaredAccessibility == Accessibility.Private)
-                        .ToImmutableDictionary(
-                            f => f.Symbol,
-                            f => f);
+            return typeDeclaration.Members

```

Comment:
I might be missing something but I think you won't handle properly partial classes/strucs (e.g. where a field is declared on some part but used on others).

Reply:
We bail out earlier when it's partial... ignore this